httpbp/client_middlewares.go

```diff
@@ -0,0 +1,194 @@
+					return counts.Requests >= config.CircuitBreaker.MinRequests && failureRatio >= config.CircuitBreaker.MinFailureRatio
+				},
+			}
+		},
+	}
+	breakers := &sync.Map{}
+
+	return func(next http.RoundTripper) http.RoundTripper {
+		return roundTripper(func(req *http.Request) (*http.Response, error) {
+			host := req.URL.Hostname()
+
+			// new circuit breakers should rarely get allocated: the pool reduces
+			// garbage collector overhead
+			newBreaker := pool.Get()
+			breaker, loaded := breakers.LoadOrStore(host, newBreaker)
+			if loaded {
+				pool.Put(newBreaker)
+			}
+
+			resp, err := breaker.(*gobreaker.CircuitBreaker).Execute(func() (interface{}, error) {
+				resp, err := next.RoundTrip(req)
+				if err != nil {
+					return nil, err
+				}
+				// circuit break on any HTTP 5xx code
+				if resp.StatusCode >= http.StatusInternalServerError {
+					err = resp.Body.Close()
+					if err != nil {
+						return nil, err
+					}
+					return nil, err
+				}
+				return resp, nil
+			})
+			if err != nil {
+				return nil, err
+			}
+			return resp.(*http.Response), nil
+		})
+	}
+}
+
+func applyMiddlewares(transport http.RoundTripper, middlewares ...ClientMiddleware) http.RoundTripper {
+	// add middlewares in reverse so the first in the list is the outermost
+	for i := len(middlewares) - 1; i >= 0; i-- {
+		transport = middlewares[i](transport)
+	}
+	return transport
+}
+
+// Retries provides a retry middleware.
+func Retries(config retry.Config) ClientMiddleware {
+	return func(next http.RoundTripper) http.RoundTripper {
+		return roundTripper(func(req *http.Request) (resp *http.Response, err error) {
+			retrybp.Do(req.Context(), func() error {
+				resp, err = next.RoundTrip(req)
+				if err != nil {
+					return err
+				}
+				return nil
+			}, retry.Attempts(1))
+			if err != nil {
+				return nil, err
+			}
+			return resp, nil
+		})
+	}
+}
+
+// LoadBalancer implements a round-robin load balancer.
+func LoadBalancer(config Config) ClientMiddleware {
+	attempts := 0
+	offset := uint64(0)
+	hosts := config.LoadBalancer.Hosts
+	return func(next http.RoundTripper) http.RoundTripper {
+		return roundTripper(func(req *http.Request) (resp *http.Response, err error) {
+
+			remaining := attempts
+			for remaining > 0 {
+				i := atomic.AddUint64(&offset, 1)
+				hostIndex := i % uint64(len(hosts))
+				req.Host = hosts[hostIndex]
+				req.URL.Host = hosts[hostIndex]
+
+				resp, err := next.RoundTrip(req)
+				if err == gobreaker.ErrOpenState || err == gobreaker.ErrTooManyRequests {
+					remaining--
+				} else {
+					return resp, err
+				}
+			}
+			return next.RoundTrip(req)
+		})
+	}
+}
+
+// MaxConcurrency is a middleware to ensure that there will be only a maximum
+// number of requests concurrently in-flight at any given time.
+func MaxConcurrency(config Config) ClientMiddleware {
+	maxConcurrency := int64(config.MaxConcurrency)
+	activeRequests := int64(0)
```

Comment:
style nit: prefer:
```suggestion
	var activeRequests int64
```
(see https://github.com/uber-go/guide/blob/master/style.md#use-var-for-zero-value-structs)